### PR TITLE
🌱 Release daily builds and manifests to staging bucket

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:
This PR adds a Makefile target that tags the latest staging images built from master branch with nightly tag and push manifests to the staging bucket. It will be called by a periodic daily GCB job (https://github.com/kubernetes/test-infra/pull/20587).

Daily images will look like this in staging registry: 
`gcr.io/k8s-staging-cluster-api/cluster-api-controller:nightly_master_20210121`
And manifest locations are: 
`https://storage.googleapis.com/artifacts.k8s-staging-cluster-api.appspot.com/components/nightly_master_20210121/bootstrap-components.yaml`

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/4006

### CODE REVIEW

**Comment by sedefsavas** [comment]
Here, new builds are created for the daily build but we can also use the latest build and tag it. One downside is a single image may have several tags if there is no new commit hence the latest build is the same for multiple daily builds. 
Anyone has a preference on this? cc @vincepri @CecileRobertMichon @randomvariable 

Master images in the staging registry are tagged as 3.10:
http://gcr.io/k8s-staging-cluster-api/cluster-api-controller
Is it possible to tag the master as v0.4.0 so that we have a differentiation from v0.3.10 release images?
I only see the last image is tagged as master, not sure where that comes from though.

File: Makefile
Comment (fabriziopandini) [inline Makefile:448]: nit: is this necessary given that we are always overriding the value in L527?

File: Makefile
Comment (fabriziopandini) [inline Makefile:527]: Might be I'm missing something but where are we setting TAG?

File: Makefile
Comment (fabriziopandini) [inline Makefile:546]: This seems a duplicate of what we have in the release target. Have you considered refactoring to avoid such duplication?

Also, I think we should add release-manifests-dev and make sure CAPD is part of the release package

**Review by fabriziopandini — COMMENTED** [review]
Thanks @sedefsavas for pushing forward this effort

Now that we are empowering our staging build process is related to out tagging process because we are usually creating tags only before the release.
E.g. before releasing v0.3.12, we create the corresponding tag
Starting from that moment, we are creating pre-release build tagged v0.3.12.something, but they should be v0.3.13.something.

What about introducing -alpha.0 tags for the next release immediately after tagging/packaging a release? this will fix also @sedefsavas problem on the v0.40 branch...

File: Makefile
Comment (sedefsavas) [reply Makefile:448]: Yes, this is used for regular release-staging. For the daily, I created a custom tag in L527.

File: Makefile
Comment (sedefsavas) [reply Makefile:532]: You're right, that is coming from Prow and set to `TAG` in cloudbuild.
Set it in Makefile like this now:

```
ifneq ($(GIT_TAG),)
  TAG ?= $(GIT_TAG)
else
  TAG ?= dev
endif
```

File: Makefile
Comment (sedefsavas) [reply Makefile:551]: Do we need CAPD images and manifests in the daily release though? CAPI e2e won't be using those.

Comment (fabriziopandini) [reply Makefile:551]: CAPD is part of the release bucket, is part of the build that is triggered at every commit, so I'm +1 to be consistent in the daily bucket as well

File: Makefile
Comment (vincepri) [inline Makefile:94]: Let's add a comment to what is this doing and why?

File: Makefile
Comment (sedefsavas) [reply Makefile]: Removed duplication and added CAPD.

**Comment by sedefsavas** [comment]
Instead of building daily images, now tagging the images that has the master tag with daily tag. By this way, we don't rebuild.

The images that has master tag in the staging registry is already built from the latest master commit.

**Comment by sedefsavas** [comment]
/hold
`GIT_TAG` may not be available to pass via `env-passthrough` from Prow job.
It is defined in GCB builds. Looking further.

**Comment by sedefsavas** [comment]
This PR is ready to go, PTAL.
After this is merged, we can merge the Prow job changes.
https://github.com/kubernetes/test-infra/pull/20587

File: Makefile
Comment (CecileRobertMichon) [inline Makefile:531]: is this comment still accurate? We're not actually building and pushing a new image anymore, right?

Comment (sedefsavas) [reply Makefile]: Correct, fixed.

**Comment by CecileRobertMichon** [comment]
/milestone v0.4.0

lgtm

/assign @nader-ziada @vincepri 


@sedefsavas can you please squash the commits?

**Comment by sedefsavas** [comment]
Squashed.

**Review by vincepri — APPROVED** [review]
/approve

Thank you for working on this @sedefsavas !

**Comment by vincepri** [comment]
@sedefsavas can you rebase?

**Comment by sedefsavas** [comment]
/hold 
checking rebase.

**Comment by sedefsavas** [comment]
/hold cancel

good to go.

**Comment by randomvariable** [comment]
lgtm too.

**Comment by sedefsavas** [comment]
I saw the test error on another PR too, this is not related to this PR.

**Pull request merged**
